### PR TITLE
Migrate pop RawQuery call sites to sqlc (proof of concept)

### PR DESCRIPTION
**Pull request by hf**
## Why

We're evaluating moving off gobuffalo/pop and onto sqlc for query generation. `RawQuery` call sites are the natural starting point: they're already hand-written SQL, so porting them doesn't require redesigning anything (unlike pop's `Eager`-loaded associations, which have no sqlc equivalent and need real query redesign). This PR is that proof of concept — every `pop.RawQuery` call in `internal/models` for sessions, refresh tokens, recovery codes, and the identity `provider_id` update now runs through sqlc-generated queries instead.

**LOC honesty check:** this diff is ~1550 lines, but only **~400 of them are hand-written and worth reviewing carefully** (the model files, `storage/dial.go`, the new `internal/models/sqlc.go` bridge, the `.sql` query files, `sqlc.yaml`, and the `Makefile`/`.gitignore` additions). The remaining **~1150 lines are 100% sqlc-generated** (`internal/storage/dbsql/sqlcgen/`, marked `DO NOT EDIT`) — that's the bulk of the diff and isn't something anyone needs to read line by line, only trust the tool for.

## The schema problem

sqlc needs to see table schemas to type its generated queries. Feeding it this repo's raw migration files doesn't work: most of them wrap their DDL in `do $$ ... end $$` blocks for idempotency (`create table if not exists`, exception handlers, etc.), and sqlc's static parser can't see into a `DO` block — it just reports the tables as never having been created. sqlc also doesn't support pointing `schema:` at a live database connection string despite that config shape looking like it should work.

The fix: `make sqlc-generate` brings up the dev postgres, runs the real migrations against it, then `pg_dump --schema-only`s the result into `internal/storage/dbsql/schema/schema.sql`, which sqlc reads. That dump is a derived artifact of the migrations, not a second source of truth, so it's gitignored and regenerated on demand rather than committed.

## The harder problem: sharing a connection between pop and sqlc

sqlc's generated code wants a `database/sql`-shaped executor (`ExecContext`/`QueryContext`/`QueryRowContext`/`PrepareContext`). A freshly-dialed `pop.Connection` satisfies that directly — its `Store` is a concrete `*pop.dB` wrapping `*sqlx.DB`, and Go promotes the full method set through concrete struct embedding.

But that's not the connection real code actually uses. Every connection that has been through `.WithContext(ctx)` or `.Transaction(...)` — which is to say, every request handler and every transactional model function — has pop rewrap `Store` in its private `contextStore` type. `contextStore` embeds the `store` type as an *interface*, not a concrete struct, and Go only promotes an embedded interface's own declared method set. That interface doesn't declare `QueryContext`/`QueryRowContext`/`PrepareContext` (pop talks to sqlx in its own vocabulary internally), so a naive type assertion silently fails on any wrapped connection.

This didn't show up as a compile error or an obvious runtime error — it showed up as a 10-minute test hang. `TestReplaceRecoveryCodesSerializesConcurrentReplacements` spawns two goroutines that both call `ReplaceRecoveryCodes` inside `db.Transaction(...)`; the first one's call failed the (silent) executor lookup, returned an error before signaling a channel, and the main goroutine sat parked on that channel forever. There's actually a comment in this codebase's existing `popConnToStd` acknowledging this exact wall ("I couldn't find any way to obtain the store when wrapped with context due to the private store field") — it just never mattered before, because that helper only ever runs once, on a freshly-dialed connection, at startup.

`storage.Connection.SQLExecutor()` (in `internal/storage/dial.go`) solves it with a bounded, `recover()`-guarded reflection unwrap that reaches past `contextStore`'s unexported field to get back to the concrete `*pop.dB`/`*pop.Tx` underneath — which, being concrete, does promote its full method set. This is the piece that makes it safe for sqlc queries to share a live pop transaction, which is what makes atomic multi-statement operations like `ReplaceRecoveryCodes` (lock + delete + insert + reset, all-or-nothing) actually work correctly.

## What's next

If we like this direction, the remaining `RawQuery` sites (webauthn challenges/credentials, factors, AMR claims, OAuth consents/authorizations, OAuth client state, a couple of test-only helpers with genuinely dynamic SQL) are mechanical follow-ups using the same pattern established here. The bigger open question for a fuller pop migration is `.Eager()`-loaded associations, which have no sqlc equivalent and need real query redesign rather than a straight port.

## Test plan
- [x] `go build ./...`
- [x] `go test ./internal/models/...` — full suite passes (was hanging indefinitely before the `SQLExecutor` fix)
- [x] `go test ./internal/storage/...` — full suite passes
- [x] `make sqlc-generate` verified end-to-end against a freshly migrated dev postgres

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by xlgmokha — APPROVED** [review]
:shipit:

File: internal/models/sqlc.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	"errors"
+
+	"github.com/supabase/auth/internal/storage"
+	"github.com/supabase/auth/internal/storage/dbsql/sqlcgen"
+)
+
+// sqlQueries returns a sqlc-generated Queries value bound to tx's current
+// executor, so statements run on the same connection or transaction as any
+// surrounding pop calls.
+func sqlQueries(tx *storage.Connection) (*sqlcgen.Queries, error) {
```
Comment (cstockton) [inline internal/models/sqlc.go:13]: I don't like this pattern at all. It steps forward the state of "tx" - an object that previously only held connection tx state, now also carries semantic meaning and forces callers to uphold an invisible ordering invariant.

File: internal/storage/dial.go [hunk 2]
```diff
@@ -68,6 +69,69 @@ func (c *Connection) Copy() *Connection {
...
+		}
+
+		next, unwrapped := unwrapFirstField(v)
+		if !unwrapped {
+			return nil, false
+		}
+		v = next
+	}
+	return nil, false
+}
```
Comment (cstockton) [inline internal/storage/dial.go:114]: I am not okay with this pattern at all. The comment says why:
```go
// SQLExecutor returns the executor backing this connection, or its current
// transaction if one is in progress
```

This removes correctness guarantees from call sites, in a way that's worst than today. I think we should move in the opposite direction where conn and tx are more or less the same thing. We don't actually know in code at any given time if we are in a transaction and write utilities / helpers to not care.

This led to all the models.Find* funcs we have today, which makes it easy to add single-purpose queries and composite them in a large number of conditional branches. I think we should move to the opposite direction, purpose built sql closest to the callers that fully leverages postgres to save round trips.

File: internal/storage/dial.go [hunk 3]
```diff
@@ -68,6 +69,69 @@ func (c *Connection) Copy() *Connection {
...
+
+// unwrapFirstField returns the value of v's first struct field, even if that
+// field is unexported (as with pop's contextStore.store). Reflection alone
+// refuses .Interface() on a value reached through an unexported field, so
+// this copies v into a freshly addressable Value and re-derives the field
+// via unsafe.Pointer, which carries no such restriction.
+func unwrapFirstField(v interface{}) (interface{}, bool) {
```
Comment (cstockton) [inline internal/storage/dial.go:121]: I think the pre-cursor work to anything is creating a new mechanism for exposing the DB, which has a reference to db without the pop abstraction. Then we don't need code like this.

File: internal/storage/dial.go [hunk 4]
```diff
@@ -68,6 +69,69 @@ func (c *Connection) Copy() *Connection {
...
+type SQLExecutor interface {
+	ExecContext(context.Context, string, ...interface{}) (sql.Result, error)
+	PrepareContext(context.Context, string) (*sql.Stmt, error)
+	QueryContext(context.Context, string, ...interface{}) (*sql.Rows, error)
+	QueryRowContext(context.Context, string, ...interface{}) *sql.Row
+}
+
+// SQLExecutor returns the executor backing this connection, or its current
+// transaction if one is in progress, for use by sqlc-generated queries.
+func (c *Connection) SQLExecutor() (executor SQLExecutor, ok bool) {
```
Comment (cstockton) [inline internal/storage/dial.go:92]: All functions that do anything db related (even if it doesn't block on network) should have a context. That way we can do observability stuff, like log / counters for recovers from panic, which should never become part of the hot paths.

File: internal/models/user.go [hunk 5]
```diff
@@ -777,11 +777,12 @@ func findUserWithLegacyRefreshToken(tx *storage.Connection, token string, forUpd
...
-		if err := tx.RawQuery(fmt.Sprintf("SELECT * FROM %q WHERE token = ? LIMIT 1 FOR UPDATE SKIP LOCKED;", refreshToken.TableName()), token).First(refreshToken); err != nil {
+		q, err := sqlQueries(tx)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+
+		if _, err := q.LockRefreshTokenForUpdate(tx.Context(), sql.NullString{String: token, Valid: true}); err != nil {
 			if errors.Cause(err) == sql.ErrNoRows {
 				return nil, nil, nil, RefreshTokenNotFoundError{}
 			}
```
Comment (cstockton) [inline internal/models/user.go:788]: This is a big loss of readability here in my eyes. I just had to really dig in to see how this works. 

What I don't like most about this implementation is something that is probably a claude specific implementation choice and not a pattern required for use of sqlc:
```
		q, err := sqlQueries(tx)
```

After looking into how this actually works, I am 100% not for this specific implementation. I would rather see code generated like:
```Go
if err := tx.RawQuery(someUniqueIdentifierForThisSpecificQuery)
```

The current magic with sqlQueries is hard to read / maintain and also dangerous in my eyes. For example imagine a func with multiple queries, later someone adds a conditional branch for an edge cases. Unfortunately it was hard to test / not covered in a unit test. Eventually this branch hits, which changes the ordering of calls to [sqlQueries](https://github.com/supabase/auth/pull/2799/changes#diff-e2f6fb3cb8ca8c78f5697bfa3a1f2cc1057873149c8ac55b6a7286d4009a8abdR13). Except the sql ordering wasn't actually updated / reflected, so the branch we wanted to avoid was executed instead, with the next function running and hitting an error since no more queries remain in `tq`.

**Review by cstockton — CHANGES_REQUESTED** [review]
I think the ORM is a big problem today for many reasons, one of the bigger reasons is I don't like the abstractions it provides. I think it encourages what we have today- where we compose many small `Find*()`, `Update*` etc  funcs that we compose together all throughout auth. This leads to many statements executed within a given auth flow, with various conditional logic or even blocking calls littered throughout.

I don't see `sqlc` materially changing that problem. To me all the work ahead is around refactoring code within `internal/api`. To select a path away from the ORM I think we need to consider how we want to approach more optimized auth flows. Many auth flows have conditional logic that comes from the database itself- I think our biggest opportunities for improvement will come from moving logic like:
* Select * from users where id = :id
* if user.field == something; update users.something
* if user.other == something; update identities.something

To simply be a purpose built function which executes that specific branch of auth server logic. Changing the multiple round trips into one. The biggest challenge with this will be refactoring surrounding code in various `internal/api` auth flows.

I think it's worth exploring beyond the orm / sql management in our software and look into the auth flows themselves, a good solution would enable taking something like:
https://github.com/supabase/auth/blob/master/internal/api/recover.go#L33
https://github.com/supabase/auth/blob/master/internal/api/mail.go#L398

Which:
* Q1: Calls models.FindUserByEmailAndAudience (outside TX)
* Q2: Calls generateFlowState (inserts into flow state) (outside TX)
* Calls sendPasswordRecovery which
    * Calls crypto.GenerateOtp
    * Calls crypto.GenerateToken
    * Sends an email (which may call postgres, smtp, etc)
    * Q3: Updates user record
    * Q4: Creates one time token (notice it inserts _after_ the email was already sent)

As we determine how to turn these 4 round trips to postgres into 1 (I think this is possible), learnings and requirements will emerge that inform us of our SQL needs. As part of that work we may find using a combination of postgres functions and stored procedures is a maintainable and high performance option, meaning we would want a solution that allows simple versioned management of them.

For example a new recover may:
* Calls otp=crypto.GenerateOtp
* Calls tok=crypto.GenerateToken
* Q1: call `v1_auth_recovery(otp, tok, params...)`
* Postgres server: v1_auth_recovery
    * Q1: Calls models.FindUserByEmailAndAudience
    * Q1: Calls generateFlowState
    * Q1: Updates user record
    * Q1: Creates one time token 
    * Q1: Insert into worker_queue (need to build it?)

Leading to a single roundtrip for the entire flow. It imposes new needs in the auth server- like a way to queue tasks for sending email. But the gains would be massive, 4 round trips and a blocking email send become 1.

There are lots of ways to go about this but my main point here is that the ORM should probably be selected after we understand the needs of a more optimized auth server.
